src/Attribute/Validator/UrlValidator.php

```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace LastCall\DownloadsPlugin\Attribute\Validator;
+
+use Composer\Package\PackageInterface;
+use LastCall\DownloadsPlugin\Attribute\AttributeManagerInterface;
+use LastCall\DownloadsPlugin\Enum\Attribute;
+
+class UrlValidator extends AbstractValidator
+{
+    public function __construct(
+        string $id,
+        PackageInterface $parent,
+        private AttributeManagerInterface $attributeManager
+    ) {
+        parent::__construct($id, $parent);
+    }
+
+    public function validate(mixed $value): string
+    {
+        if (null === $value) {
+            $this->throwException($this->getAttribute(), 'is required');
+        }
+
+        if (!\is_string($value)) {
+            $this->throwException($this->getAttribute(), sprintf('must be string, "%s" given', get_debug_type($value)));
+        }
+
+        $url = strtr($value, $this->attributeManager->get(Attribute::VARIABLES));
+        if (false === filter_var($url, \FILTER_VALIDATE_URL)) {
+            $this->throwException($this->getAttribute(), 'is invalid url');
+        }
+
+        if (!\in_array(parse_url($url, \PHP_URL_SCHEME), ['http', 'https'])) {
+            $this->throwException($this->getAttribute(), 'has invalid scheme');
+        }
```

Comment:
What about:

- Handling local URLs starting with `file://` (particularly important if this is being run in an isolated environment without access to the internet)

- Could it be worth delegated the schemes to whatever the underlying library supports? In case, for example, the library also supports `nfs://` or `ftp://`?

Reply:
This is a good question.

* I think the original forks allow these URLs
* I tested and Composer support `file://` and `ftp://`. I didn't try `nfs://`, but I think it support also.
* At the time I restrict to `http://` and `https://`, I was worry about something like this `file:///etc/passwd`. I was worry that a library will use this plugin to download a local file and put it somewhere for a purpose. Probably I was worry too much? The library can just call `file_get_contents('/etc/passwd')`
* I didn't think about `ftp://` and `nfs://` at the time I made the restriction.
* Is it a good idea to just remove this restriction, just like the original forks?

Reply:
I think that leaving the choices open is fine. If ultimately Composer doesn't support a URL, then the end-user will know very quickly anyway.

Regarding a possible security risk, while I admit it is there, the very fact that we are pulling files from elsewhere is a risk. I don't think it is inherently more risky to have a `file://` URI. And as you rightly mention, one could just do a `file_get_contents('/etc/passwd')`.

Reply:
Create separate PR for this change https://github.com/pact-foundation/composer-downloads-plugin/pull/33